packages/react-menu/src/components/MenuItemCheckbox/useMenuItemCheckbox.ts

```diff
@@ -0,0 +1,46 @@
+import * as React from 'react';
+import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utils';
+import { MenuItemCheckboxProps, MenuItemCheckboxState } from './MenuItemCheckbox.types';
+import { useMenuItemSelectable } from '../../selectable/index';
+import { useMergedRefs } from '@fluentui/react-hooks';
+
+/**
+ * Consts listing which props are shorthand props.
+ */
+export const menuItemCheckboxShorthandProps = ['icon', 'checkmark'];
+
+const mergeProps = makeMergeProps<MenuItemCheckboxState>({ deepMerge: menuItemCheckboxShorthandProps });
+
+/** Returns the props and state required to render the component */
+export const useMenuItemCheckbox = (
+  props: MenuItemCheckboxProps,
+  ref: React.Ref<HTMLElement>,
+  defaultProps?: MenuItemCheckboxProps,
+): MenuItemCheckboxState => {
+  const state = mergeProps(
+    {
+      ref: useMergedRefs(ref, React.useRef(null)),
+      as: 'div',
+      icon: { as: 'span' },
+      checkmark: { as: 'span' },
+      role: 'menuitemcheckbox',
+      tabIndex: 0,
+    },
+    defaultProps,
+    resolveShorthandProps(props, menuItemCheckboxShorthandProps),
+  );
+
+  useMenuItemSelectable(state, () => {
+    const newCheckedItems = [...state.checkedItems];
+    const index = state.checkedItems.indexOf(state.value);
+    if (index !== -1) {
+      newCheckedItems.splice(index, 1);
+    } else {
+      newCheckedItems.push(state.value);
+    }
+
+    return newCheckedItems;
+  });
```

Comment:
This looks confusing, let's how it will fit in the whole picture of `Menu`'s state. IMO, we can use the similar approach as in `useTree()` where parents holds and manages the state.

https://github.com/microsoft/fluentui/blob/7566015a7edd355b4fcd3796bc8f44f732ef0877/packages/fluentui/react-northstar/src/components/Tree/hooks/useTree.ts#L246

Reply:
Yeah let's see how the the implementation comes together with context selector, I'll make a not to address it